networkx/convert_matrix.py

```diff
@@ -121,51 +121,103 @@ def to_pandas_dataframe(G, nodelist=None, multigraph_weight=sum, weight='weight'
+      directed case).
 
-    Notes
-    -----
-    If the numpy matrix has a single data type for each matrix entry it
-    will be converted to an appropriate Python data type.
+    target : str (or int)
+      Column name (or index if no column names) of the target nodes (for the
+      directed case).
 
-    If the numpy matrix has a user-specified compound data type the names
-    of the data fields will be used as attribute keys in the resulting
-    NetworkX graph.
+    edge_attr : str (or int), iterable, True
+      A name (str or index if no column names) or list of names that will be
+      used to retrieve attributes from the row that should be added to the
+      graph as edge attributes. If `True` is passed as the key, all of the
+      remaining columns will be added.
+
+    create_using : NetworkX graph
+       Use specified graph for result.  The default is Graph()
 
     See Also
     --------
-    from_pandas_dataframe
+    to_pandas_dataframe
 
     Examples
     --------
     Simple integer weights on edges:
 
     >>> import pandas as pd
-    >>> df=pd.DataFrame([[1,1],[2,1]])
-    >>> G=nx.from_pandas_dataframe(df)
+    >>> import numpy as np
+    >>> r = np.random.RandomState(seed=5)
+    >>> ints = r.random_integers(1, 10, size=(3,2))
+    >>> a = ['A', 'B', 'C']
+    >>> b = ['D', 'A', 'E']
+    >>> df = pd.DataFrame(ints, columns=['weight', 'cost'])
+    >>> df
+       weight  cost  0  b
+    0       4     7  A  D
+    1       7     1  B  A
+    2      10     9  C  E
+    >>> df[0] = a
+    >>> df['b'] = b
+    >>> G=nx.from_pandas_dataframe(df, 0, 'b', ['weight', 'cost'])
+    >>> G.edges(data=True)
+    [('E','C',{'cost':9,'weight':10}),
+    ('B','A',{'cost':1,'weight':7}),
+    ('A','D',{'cost':7,'weight':4})]
     """
 
-    import pandas as pd
-    A = df.values
-    G = from_numpy_matrix(A, create_using)
-    try:
-        df = df[df.index]
-    except:
-        raise nx.NetworkXError("Columns must match Indices.",
-                               "%s not in columns"%list(set(df.index).difference(set(df.columns))))
-    nx.relabel.relabel_nodes(G, dict(enumerate(df.columns)), copy=False)
-    return G
+    g = _prep_create_using(create_using)
+
+    if edge_attr:
+        # If all additional columns requested
+        if edge_attr == True:
+            # Create a list of all columns
+            cols = list(df.columns)
+            cols.remove(source)
+            cols.remove(target)
+            edge_attr = cols
+        # If a list or tuple of name is requested
+        elif isinstance(edge_attr, (list, tuple)):
+            edge_attr = list(edge_attr)
+        # If a string or int (index) is passed
+        else:
+            edge_attr = [edge_attr,]
+
+        length = len(edge_attr)
+        if length > 1:
+            for idx, row in df.iterrows():
+                g.add_edge(row[source], row[target], **row[edge_attr])
+        # If only one edge_attr, can't use kwarg expansion
```

Comment:
Since we still made `edge_attr` a list above, we can actually use kwarg expansion.